File: lib/go/transport.go
package frugal

import (
	"bytes"
	"encoding/binary"
	//"errors"

	"git.apache.org/thrift.git/lib/go/thrift"
)

// FPublisherTransportFactory produces FPublisherTransports and is typically
// used by an FScopeProvider.
type FPublisherTransportFactory interface {
	GetTransport() FPublisherTransport
}

// FPublisherTransport is used exclusively for pub/sub scopes. Publishers use it
// to publish messages to a topic.
type FPublisherTransport interface {
	// Open opens the transport.
	Open() error

	// Close closes the transport.
	Close() error

	// IsOpen returns true if the transport is open, false otherwise.
	IsOpen() bool

	// GetPublishSizeLimit returns the maximum allowable size of a payload
	// to be published. A non-positive number is returned to indicate an
	// unbounded allowable size.
	GetPublishSizeLimit() uint

	// Publish sends the given payload with the transport. Implementations
	// of publish should be threadsafe.
	Publish(string, []byte) error
}

// FSubscriberTransportFactory produces FSubscriberTransports and is typically
// used by an FScopeProvider.
type FSubscriberTransportFactory interface {
	GetTransport() FSubscriberTransport
}

// FSubscriberTransport is used exclusively for pub/sub scopes. Subscribers use
// it to subscribe to a pub/sub topic.
type FSubscriberTransport interface {
	// Subscribe opens the transport and sets the subscribe topic.
	Subscribe(string, FAsyncCallback) error

	// Unsubscribe unsubscribes from the topic and closes the transport.
	Unsubscribe() error

	// IsSubscribed returns true if the transport is subscribed to a topic,
	// false otherwise.
	IsSubscribed() bool
}

type FUnifiedPublisherTransportFactory interface {
	GetTransport() FUnifiedPublisherTransport
}

type FUnifiedPublisherTransport interface {
	Open() error

	Close() error

	IsOpen() bool

	GetPublishSizeLimit() uint

	PublishEphemeral(string, []byte) error

	PublishDurable(string, *string, []byte) error
}

type FUnifiedSubscriberTransportFactory interface {
	GetTransport() FUnifiedSubscriberTransport
}

type FUnifiedSubscriberTransport interface {
	Subscribe(string, FAsyncCallback) error

	Unsubscribe() error

	IsSubscribed() bool
}

// FDurablePublisherTransportFactory produces FDurablePublisherTransports and is
// typically used by an FDurableScopeProvider.
type FDurablePublisherTransportFactory interface {
	GetTransport() FDurablePublisherTransport
}

// FDurableSubscriberTransportFactory produces FDurableSubscriberTransports and
// is typically used by an FDurableScopeProvider.
type FDurableSubscriberTransportFactory interface {
	GetTransport() FDurableSubscriberTransport
}

// FDurableAsyncCallback is an internal callback which is constructed by
// generated code and invoked by an FDurableSubscriberTransport when a message
// is received.
type FDurableAsyncCallback func(thrift.TTransport, *string) error

// FDurablePublisherTransport is used exclusively for pub/sub scopes. Publishers
// use it to publish messages durably to a topic.
type FDurablePublisherTransport interface {
	// Open opens the transport.
	Open() error

	// Close closes the transport.
	Close() error

	// IsOpen returns true if the transport is open, false otherwise.
	IsOpen() bool

	// GetPublishSizeLimit returns the maximum allowable size of a payload
	// to be published. A non-positive number is returned to indicate an
	// unbounded allowable size.
	GetPublishSizeLimit() uint

	// Publish sends the given payload with the transport. Implementations
	// of publish should be threadsafe. The parameter groupID allows you
	// to assign arbitrary groupings to messages. A value of nil indicates
	// the message isn't part of any group.
	Publish(subejct string, groupID *string, payload []byte) error
}

// FDurableSubscriberTransport is used exclusively for pub/sub scopes.
// Subscribers use it to durably subscribe to a pub/sub topic.
type FDurableSubscriberTransport interface {
	// Subscribe opens the transport and sets the subscribe topic.
	Subscribe(string, FDurableAsyncCallback) error

	// Unsubscribe unsubscribes from the topic and closes the transport.
	Unsubscribe() error

	// IsSubscribed returns true if the transport is subscribed to a topic,
	// false otherwise.
	IsSubscribed() bool
}

// FTransport is Frugal's equivalent of Thrift's TTransport. FTransport is
// comparable to Thrift's TTransport in that it represents the transport layer
// for frugal clients. However, frugal is callback based and sends only framed
// data. Due to this, instead of read, write, and flush methods, FTransport has
// a send method that sends framed frugal messages. To handle callback data, an
// FTransport also has an FRegistry, so it provides methods for registering
// and unregistering an FAsyncCallback to an FContext.
type FTransport interface {
	// SetMonitor starts a monitor that can watch the health of, and reopen,
	// the transport.
	SetMonitor(FTransportMonitor)

	// Closed channel receives the cause of an FTransport close (nil if clean
	// close).
	Closed() <-chan error

	// Open prepares the transport to send data.
	Open() error

	// IsOpen returns true if the transport is open, false otherwise.
	IsOpen() bool

	// Close closes the transport.
	Close() error

	// Oneway transmits the given data and doesn't wait for a response.
	// Implementations of oneway should be threadsafe and respect the timeout
	// present on the context.
	Oneway(ctx FContext, payload []byte) error

	// Request transmits the given data and waits for a response.
	// Implementations of request should be threadsafe and respect the timeout
	// present on the context.
	Request(ctx FContext, payload []byte) (thrift.TTransport, error)

	// GetRequestSizeLimit returns the maximum number of bytes that can be
	// transmitted. Returns a non-positive number to indicate an unbounded
	// allowable size.
	GetRequestSizeLimit() uint
}

// FTransportFactory produces FTransports by wrapping a provided TTransport.
type FTransportFactory interface {
	GetTransport(tr thrift.TTransport) FTransport
}

type fBaseTransport struct {
	requestSizeLimit uint
	writeBuffer      bytes.Buffer
	registry         fRegistry
	closed           chan error
}

// Initialize a new fBaseTransport
func newFBaseTransport(requestSizeLimit uint) *fBaseTransport {
	return &fBaseTransport{
		requestSizeLimit: requestSizeLimit,
		registry:         newFRegistry(),
	}
}

// Intialize the close channels
func (f *fBaseTransport) Open() {
	f.closed = make(chan error, 1)
}

// Close the close channels
func (f *fBaseTransport) Close(cause error) {
	select {
	case f.closed <- cause:
	default:
		logger().Warnf("frugal: unable to put close error '%s' on fBaseTransport closed channel", cause)
	}
	close(f.closed)
}

// Execute a frugal frame (NOTE: this frame must include the frame size).
func (f *fBaseTransport) ExecuteFrame(frame []byte) error {
	return f.registry.Execute(frame[4:])
}

// Closed channel is closed when the FTransport is closed.
func (f *fBaseTransport) Closed() <-chan error {
	return f.closed
}

func prependFrameSize(buf []byte) []byte {
	frame := make([]byte, 4)
	binary.BigEndian.PutUint32(frame, uint32(len(buf)))
	return append(frame, buf...)
}

File: lib/python/frugal/transport/transport.py
# Copyright 2017 Workiva
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#     http://www.apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from thrift.transport.TTransport import TTransportBase
from thrift.transport.TTransport import TTransportException

from frugal.exceptions import TTransportExceptionType


class FTransport(object):
    """
    FTransport is comparable to Thrift's TTransport in that it represent the
    transport layer for frugal clients. However, frugal is callback based and
    sends only framed data. Therefore, instead of exposing read, write, and
    flush, the transport has a simple request method that sends framed frugal
    messages and returns the response.
    """
    def __init__(self, request_size_limit=0):
        """
        Initialize the transport optionally passing a request size limit.

        Args:
            request_size_limit: The maximum request payload size for this
                                transport. A non-positive number indicates
                                an unbounded allowable size.
        """
        self._request_size_limit = request_size_limit

    def open(self):
        """
        Open the transport.
        """
        raise NotImplementedError('You must override this')

    def close(self):
        """
        Close the transport.
        """
        raise NotImplementedError('You must override this')

    def is_open(self):
        """
        Return True if the transport is open, False otherwise.
        """
        raise NotImplementedError('You must override this')

    def set_monitor(self, monitor):
        """
        Set the transport monitor for the transport. This should only be used
        for "stateful" transports.

        Args:
            monitor: A transport monitor
        """
        raise NotImplementedError('You must override this')

    def oneway(self, context, payload):
        """
        Send the given framed frugal payload over the transport.

        Args:
            context:    FContext associated with the request (used for timeout
                        and logging)

            payload:    framed frugal data
        """
        raise NotImplementedError('You must override this')

    def request(self, context, payload):
        """
        Send the given framed frugal payload over the transport and returns the
        response.

        Args:
            context:    FContext associated with the request (used for timeout
                        and logging)

            payload:    framed frugal data
        """
        raise NotImplementedError('You must override this')

    def get_request_size_limit(self):
        """
        Returns the maximum request payload size for this transport. A non-
        positive number is returned to indicate an unbounded allowable size.
        """
        return self._request_size_limit

    def _preflight_request_check(self, payload):
        """
        Helper function that throws TTransportExceptionType.NOT_OPEN if the
        transport is not open or throws FMessageSizeException if the payload is
        too large. Should only be called by extending classes.
        """
        if not self.is_open():
            raise TTransportException(
                type=TTransportExceptionType.NOT_OPEN,
                message='Transport is not open')

        if len(payload) > self.get_request_size_limit() > 0:
            raise TTransportException(
                type=TTransportExceptionType.REQUEST_TOO_LARGE,
                message='Message exceeds {0} bytes, was {1} bytes'.format(
                    self.get_request_size_limit(), len(payload)))


class TSynchronousTransport(TTransportBase, object):
    """
    TSynchronousTransport is a Thrift TTransport for services which make
    synchronous requests.
    """

    def set_timeout(self, timeout):
        """
        Set the request timeout.

        Args:
            timeout: request timeout in milliseconds.
        """
        pass
